File: src/9장/9-4-1.py
import sys
sys.setrecursionlimit(20000)
input = sys.stdin.readline

class Node:
    def __init__(self,data):
        self.data = data
        self.left = None
        self.right = None

class Tree:
    def __init__(self):
        self.root = None

    def add(self,data):
        if(self.root == None):
            self.root = Node(data)
        
        else:
            current = self.root
            while(True):
                if(current.data > data):
                    if(current.left == None):
                        current.left = Node(data)
                        break
                    current = current.left

                if(current.data < data):
                    if(current.right == None):
                        current.right = Node(data)
                        break
                    current = current.right
    
    def postOrder(self,node=None):
        global answer
        
        if node == None:
            node = self.root
        
        if node.left != None:
            self.postOrder(node.left)
        if node.right != None:
            self.postOrder(node.right)
        answer.append(node.data)


tree = Tree()
while True:
    try:
        tree.add(int(input()))
    except:
        break
answer = []
tree.postOrder()
print('\n'.join(map(str,answer)))

File: src/9장/9-4-1.java
import java.util.*;
import java.io.*;

public class Main {
    static class Node {
        int data;
        Node left, right;
        Node(int data) {
            this.data = data;
        }
        Node(int data, Node left, Node right) {
            this.data = data;
            this.left = left;
            this.right = right;
        }
        void insert(int n) {
            if (n < this.data) {
                if (this.left == null)
                    this.left = new Node(n);
                else this.left.insert(n);
            } else {
                if (this.right == null)
                    this.right = new Node(n);
                else this.right.insert(n);
            }
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        Node tree = new Node(Integer.parseInt(br.readLine()));
        String input;
        while (true) {
            input = br.readLine();
            if (input == null || input.equals(""))
                break;
				tree.insert(Integer.parseInt(input));
        }
        postOrder(tree);
    }

    static void postOrder(Node tree) {
        if (tree == null)return;
        postOrder(tree.left);
        postOrder(tree.right);
        System.out.println(tree.data);
    }
}